cpp/src/arrow/util/small_vector.h

```diff
@@ -0,0 +1,509 @@
+  using iterator_category = std::random_access_iterator_tag;
+
+  // Some algorithms need to default-construct an iterator
+  constexpr VectorIterator() noexcept = default;
+
+  constexpr explicit VectorIterator(pointer ptr) noexcept : ptr_(ptr) {}
+
+  // Value access
+  constexpr reference operator*() const { return *ptr_; }
+
+  constexpr reference operator[](difference_type n) const { return ptr_[n]; }
+
+  // Forward / backward
+  VectorIterator& operator++() {
+    ++ptr_;
+    return *this;
+  }
+  VectorIterator& operator--() {
+    --ptr_;
+    return *this;
+  }
+  VectorIterator operator++(int) {
+    VectorIterator tmp(*this);
+    ++ptr_;
+    return tmp;
+  }
+  VectorIterator operator--(int) {
+    VectorIterator tmp(*this);
+    --ptr_;
+    return tmp;
+  }
+
+  // Arithmetic
+  constexpr difference_type operator-(const VectorIterator& other) const {
+    return ptr_ - other.ptr_;
+  }
+  constexpr VectorIterator operator+(difference_type n) const {
+    return VectorIterator(ptr_ + n);
+  }
+  constexpr VectorIterator operator-(difference_type n) const {
+    return VectorIterator(ptr_ - n);
+  }
+  VectorIterator& operator+=(difference_type n) {
+    ptr_ += n;
+    return *this;
+  }
+  VectorIterator& operator-=(difference_type n) {
+    ptr_ -= n;
+    return *this;
+  }
+
+  // Comparisons
+  constexpr bool operator==(const VectorIterator& other) const {
+    return ptr_ == other.ptr_;
+  }
+  constexpr bool operator!=(const VectorIterator& other) const {
+    return ptr_ != other.ptr_;
+  }
+  constexpr bool operator<(const VectorIterator& other) const {
+    return ptr_ < other.ptr_;
+  }
+  constexpr bool operator>(const VectorIterator& other) const {
+    return ptr_ > other.ptr_;
+  }
+  constexpr bool operator<=(const VectorIterator& other) const {
+    return ptr_ <= other.ptr_;
+  }
+  constexpr bool operator>=(const VectorIterator& other) const {
+    return ptr_ >= other.ptr_;
+  }
+
+ private:
+  pointer ptr_ = NULLPTR;
+};
+
+template <typename T>
+class StaticVectorMixin {
+ protected:
+  // properly aligned uninitialized storage for N T's
+  using storage_type = typename std::aligned_storage<sizeof(T), alignof(T)>::type;
+
+  static T* ptr_at(storage_type* p, size_t i) {
+    return launder(reinterpret_cast<T*>(&p[i]));
+  }
+
+  static const T* ptr_at(const storage_type* p, size_t i) {
+    return launder(reinterpret_cast<const T*>(&p[i]));
+  }
+
+  static void move_storage(storage_type* src, storage_type* dest, size_t n) {
+    for (size_t i = 0; i < n; ++i) {
+      T* src_item = ptr_at(src, i);
+      T* dest_item = ptr_at(dest, i);
+      new (dest_item) T(std::move(*src_item));
+      src_item->~T();
+    }
+  }
+
+  static void destroy_storage(storage_type* p, size_t n) {
+    for (size_t i = 0; i < n; ++i) {
+      ptr_at(p, i)->~T();
+    }
+  }
+};
+
+template <typename T, size_t N, bool NonTrivialDestructor>
+class StaticVectorBaseStorage : public StaticVectorMixin<T> {
```

Comment:
Could we avoid inheritance here?

Reply:
How do you suggest to avoid it? We can use composition but it will only add some indirection boilerplate.

Reply:
I prefer some indirection boilerplate to inheritance in this case; I made a sketch of what that'd look like in the comment below.

Reply:
Do you still want me to try composition?

Reply:
Note one question is how to share the definition of most vector methods between SmallVector and StaticVector. Your sketch doesn't show that.

Reply:
I would like to see this use composition instead. Honestly, I think in this case since there are only two classes it'd be better to keep the two unrelated and repeat implementations (modulo calls to the helper trait). I can make a meta-PR to your branch tomorrow with a more thorough sketch if you'd prefer